# Turbofan Engines Predictive Maintenance

## EDA Phase

""" Aircraft Turbofan Engines Predictive Maintenance:
This notebook is the EDA phase for 11 turbofan Engines,
the ultimate goal of the complete project is to predict
the remaining useful life (RUL) of similar in-service engines """


In [2]:
# Import neccessary libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


In [3]:
# load the data
df = pd.read_csv('engines data 130325 masked.csv') 


In [4]:
df.head(20)

EPOSITION FLIGHT_PHASE FLIGHT_DATETIME ENGINE_INSTALL_DATE  Cycles  \
0           1       CRUISE   29-11-23 8:56      24-08-23 23:00   334.0   
1           1       CRUISE   07-09-23 1:10      24-08-23 23:00    32.0   
2           1       CRUISE   19-01-24 3:09      24-08-23 23:00   532.0   
3           1        CLIMB  25-02-24 21:58      24-08-23 23:00   675.0   
4           1      TAKEOFF  10-07-24 16:28      24-08-23 23:00  1223.0   
5           1       CRUISE  23-02-25 23:45      24-08-23 23:00   126.0   
6           1       CRUISE   17-03-24 8:01      24-08-23 23:00   766.0   
7           1       CRUISE  08-12-24 12:53      24-08-23 23:00   118.0   
8           1      TAKEOFF   09-08-24 2:11      24-08-23 23:00    39.0   
9           1      TAKEOFF   30-06-24 9:25      24-08-23 23:00  1178.0   
10          1       CRUISE   03-10-23 3:45      24-08-23 23:00   126.0   
11          1      TAKEOFF   04-02-24 9:02      24-08-23 23:00   594.0   
12          1       CRUISE   19-12-24 8:32      24-08-23 23:00   154.0   
13          1        CLIMB  30-03-24 12:48      24-08-23 23:00   819.0   
14          1        CLIMB  05-06-24 14:13      24-08-23 23:00  1086.0   
15          1       CRUISE   01-12-24 2:30      24-08-23 23:00    96.0   
16          1       CRUISE   05-01-25 8:34      24-08-23 23:00   181.0   
17          1        CLIMB  30-11-24 13:51      24-08-23 23:00    94.0   
18          1       CRUISE   30-11-24 7:43      24-08-23 23:00    93.0   
19          1       CRUISE   12-01-24 5:08      24-08-23 23:00   504.0   

    EGT Probe average  Fuel Flow  Corrected Fan Speed  Core Speed  ZPN12P  \
0            793.3125     2782.0              92.2129      103.12   86.67   
1            792.9250     2725.0              90.0819      103.99   86.02   
2            757.2250     2861.0              89.0560      101.87   83.67   
3            905.0375     5959.0                  NaN      107.42   92.15   
4            912.3625     2911.0              85.3176      106.87   86.45   
5            806.4375     2650.0              92.0863      103.02   86.50   
6            715.6875     2504.0              88.0891       99.87   81.40   
7            786.9625     3005.0              92.2527      102.16   85.80   
8            983.2375     4066.0              86.9904      110.24   90.41   
9            938.3250     3206.0              85.0595      108.41   87.59   
10           801.7250     2964.0              89.9929      104.24   86.18   
11           883.4250     1917.0              84.8623      105.44   84.98   
12           792.7625     2851.0              91.5343      102.63   85.88   
13           880.2750     5963.0                  NaN      106.23   91.36   
14           848.1250     5795.0                  NaN      104.89   89.37   
15           790.1250     2910.0              89.2138      103.38   85.11   
16           737.3125     2801.0              89.1492      100.30   82.48   
17           880.3625     5631.0                  NaN      105.81   90.54   
18           774.6000     3028.0              91.0355      102.02   84.99   
19           767.9000     2733.0              88.4549      102.49   83.74   

    Vibration N1 #1 Bearing  Vibration N2 #1 Bearing  \
0                    0.1267                    0.500   
1                    0.1200                    0.200   
2                    0.0933                    0.350   
3                    0.1133                    0.450   
4                    0.0600                    0.850   
5                    0.0000                    0.700   
6                    0.0400                    0.125   
7                    0.0000                    0.500   
8                    0.0933                    0.400   
9                    0.0667                    0.650   
10                   0.1800                    0.650   
11                   0.0667                    0.125   
12                   0.0400                    0.375   
13                   0.1200                    

In [5]:
# Check the data statistics
df.describe()

EPOSITION         Cycles  EGT Probe average      Fuel Flow  \
count  120677.000000  120672.000000      120610.000000  120603.000000   
mean        1.513594    1919.581867         825.060648    3306.745645   
std         0.499817    1388.496433          80.575206    1548.253271   
min         1.000000       0.000000           0.000000     553.000000   
25%         1.000000     682.000000         763.312500    2168.000000   
50%         2.000000    1693.000000         849.937500    2704.000000   
75%         2.000000    3076.000000         883.662500    5107.000000   
max         2.000000    5003.000000        1047.600000    9925.000000   

       Corrected Fan Speed     Core Speed         ZPN12P  \
count         79972.000000  120643.000000  120643.000000   
mean             84.621995     103.296047      83.395046   
std               5.013564       4.532367       8.490425   
min              68.178800       0.820000       0.240000   
25%              79.768825     101.410000      81.060000   
50%              85.368350     104.280000      83.770000   
75%              89.027550     106.040000      88.330000   
max             100.156900     112.130000      95.310000   

       Vibration N1 #1 Bearing  Vibration N2 #1 Bearing  \
count            120379.000000            120377.000000   
mean                  0.134921                 0.339130   
std                   0.097757                 0.240131   
min                   0.000000                 0.000000   
25%                   0.060000                 0.150000   
50%                   0.113300                 0.300000   
75%                   0.200000                 0.475000   
max                   0.793300                 1.600000   

       Vibration N2 Turbine Frame    Engine Type  Engine Number  
count               120379.000000  120677.000000  120677.000000  
mean                     0.597473       1.782005       5.172319  
std                      0.392315       0.412886       2.938684  
min                      0.000000       1.000000       1.000000  
25%                      0.325000       2.000000       3.000000  
50%                      0.500000       2.000000       5.000000  
75%                      0.775000       2.000000       7.000000  
max                      2.525000       2.000000      11.000000

In [6]:
# check data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120677 entries, 0 to 120676
Data columns (total 15 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   EPOSITION                   120677 non-null  int64  
 1   FLIGHT_PHASE                120677 non-null  object 
 2   FLIGHT_DATETIME             120677 non-null  object 
 3   ENGINE_INSTALL_DATE         120677 non-null  object 
 4   Cycles                      120672 non-null  float64
 5   EGT Probe average           120610 non-null  float64
 6   Fuel Flow                   120603 non-null  float64
 7   Corrected Fan Speed         79972 non-null   float64
 8   Core Speed                  120643 non-null  float64
 9   ZPN12P                      120643 non-null  float64
 10  Vibration N1 #1 Bearing     120379 non-null  float64
 11  Vibration N2 #1 Bearing     120377 non-null  float64
 12  Vibration N2 Turbine Frame  120379 non-null  float64
 13  Engine Type   

In [7]:
# Check for missing values, percentage
df.isnull().sum()/len(df)*100


EPOSITION                      0.000000
FLIGHT_PHASE                   0.000000
FLIGHT_DATETIME                0.000000
ENGINE_INSTALL_DATE            0.000000
Cycles                         0.004143
EGT Probe average              0.055520
Fuel Flow                      0.061321
Corrected Fan Speed           33.730537
Core Speed                     0.028174
ZPN12P                         0.028174
Vibration N1 #1 Bearing        0.246940
Vibration N2 #1 Bearing        0.248597
Vibration N2 Turbine Frame     0.246940
Engine Type                    0.000000
Engine Number                  0.000000
dtype: float64

In [8]:
# delete features that have more than 50% missing values
null_50 = df.isnull().sum()/len(df)*100 > 50
null_50 = null_50[null_50].index
df = df.drop(null_50, axis=1)
df.head(20)
 

EPOSITION FLIGHT_PHASE FLIGHT_DATETIME ENGINE_INSTALL_DATE  Cycles  \
0           1       CRUISE   29-11-23 8:56      24-08-23 23:00   334.0   
1           1       CRUISE   07-09-23 1:10      24-08-23 23:00    32.0   
2           1       CRUISE   19-01-24 3:09      24-08-23 23:00   532.0   
3           1        CLIMB  25-02-24 21:58      24-08-23 23:00   675.0   
4           1      TAKEOFF  10-07-24 16:28      24-08-23 23:00  1223.0   
5           1       CRUISE  23-02-25 23:45      24-08-23 23:00   126.0   
6           1       CRUISE   17-03-24 8:01      24-08-23 23:00   766.0   
7           1       CRUISE  08-12-24 12:53      24-08-23 23:00   118.0   
8           1      TAKEOFF   09-08-24 2:11      24-08-23 23:00    39.0   
9           1      TAKEOFF   30-06-24 9:25      24-08-23 23:00  1178.0   
10          1       CRUISE   03-10-23 3:45      24-08-23 23:00   126.0   
11          1      TAKEOFF   04-02-24 9:02      24-08-23 23:00   594.0   
12          1       CRUISE   19-12-24 8:32      24-08-23 23:00   154.0   
13          1        CLIMB  30-03-24 12:48      24-08-23 23:00   819.0   
14          1        CLIMB  05-06-24 14:13      24-08-23 23:00  1086.0   
15          1       CRUISE   01-12-24 2:30      24-08-23 23:00    96.0   
16          1       CRUISE   05-01-25 8:34      24-08-23 23:00   181.0   
17          1        CLIMB  30-11-24 13:51      24-08-23 23:00    94.0   
18          1       CRUISE   30-11-24 7:43      24-08-23 23:00    93.0   
19          1       CRUISE   12-01-24 5:08      24-08-23 23:00   504.0   

    EGT Probe average  Fuel Flow  Corrected Fan Speed  Core Speed  ZPN12P  \
0            793.3125     2782.0              92.2129      103.12   86.67   
1            792.9250     2725.0              90.0819      103.99   86.02   
2            757.2250     2861.0              89.0560      101.87   83.67   
3            905.0375     5959.0                  NaN      107.42   92.15   
4            912.3625     2911.0              85.3176      106.87   86.45   
5            806.4375     2650.0              92.0863      103.02   86.50   
6            715.6875     2504.0              88.0891       99.87   81.40   
7            786.9625     3005.0              92.2527      102.16   85.80   
8            983.2375     4066.0              86.9904      110.24   90.41   
9            938.3250     3206.0              85.0595      108.41   87.59   
10           801.7250     2964.0              89.9929      104.24   86.18   
11           883.4250     1917.0              84.8623      105.44   84.98   
12           792.7625     2851.0              91.5343      102.63   85.88   
13           880.2750     5963.0                  NaN      106.23   91.36   
14           848.1250     5795.0                  NaN      104.89   89.37   
15           790.1250     2910.0              89.2138      103.38   85.11   
16           737.3125     2801.0              89.1492      100.30   82.48   
17           880.3625     5631.0                  NaN      105.81   90.54   
18           774.6000     3028.0              91.0355      102.02   84.99   
19           767.9000     2733.0              88.4549      102.49   83.74   

    Vibration N1 #1 Bearing  Vibration N2 #1 Bearing  \
0                    0.1267                    0.500   
1                    0.1200                    0.200   
2                    0.0933                    0.350   
3                    0.1133                    0.450   
4                    0.0600                    0.850   
5                    0.0000                    0.700   
6                    0.0400                    0.125   
7                    0.0000                    0.500   
8                    0.0933                    0.400   
9                    0.0667                    0.650   
10                   0.1800                    0.650   
11                   0.0667                    0.125   
12                   0.0400                    0.375   
13                   0.1200                    

In [9]:
# lets rename the columns for easy access, replace space with underscore, and convert to lower case, and remove special characters, and remove the word 'engine', and make shorter names
df.columns = df.columns.str.replace(' ', '_').str.lower().str.replace('[-/()#]', '').str.replace('engine', 'eng').str.replace('temperature', 'temp').str.replace('pressure', 'pres').str.replace('vibration', 'vib').str.replace('speed', 'spd').str.replace('temperature', 'temp').str.replace('flow', 'flw')

In [10]:
df.columns

Index(['eposition', 'flight_phase', 'flight_datetime', 'eng_install_date',
       'cycles', 'egt_probe_average', 'fuel_flw', 'corrected_fan_spd',
       'core_spd', 'zpn12p', 'vib_n1_#1_bearing', 'vib_n2_#1_bearing',
       'vib_n2_turbine_frame', 'eng_type', 'eng_number'],
      dtype='object')

In [11]:
# check how many engines in the data
df['eng_number'].unique()

array([11,  8,  6,  7,  2,  5,  3, 10,  9,  1,  4], dtype=int64)

In [12]:
# make a dictionary of engines types and numbers for each type
eng_types = df['eng_type'].unique()
eng_numbers = {}
for eng in eng_types:
    eng_numbers[eng] = df[df['eng_type'] == eng]['eng_number'].unique()
eng_numbers


{1: array([11,  8, 10,  9], dtype=int64),
 2: array([6, 7, 2, 5, 3, 1, 4], dtype=int64)}

In [13]:
# Change the categorical dat types into suitable type

df['eng_number'] = df['eng_number'].astype('category')
df['eng_type'] = df['eng_type'].astype('category')
df['flight_phase'] = df['flight_phase'].astype('category')


In [14]:
# check the unique flight phases
df['flight_phase'].unique()

['CRUISE', 'CLIMB', 'TAKEOFF']
Categories (3, object): ['CLIMB', 'CRUISE', 'TAKEOFF']

In [15]:
# how many nulls in each flight phase
df.groupby('flight_phase').count()

C:\Users\Ahmad\AppData\Local\Temp\ipykernel_14048\1453023314.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('flight_phase').count()


eposition  flight_datetime  eng_install_date  cycles  \
flight_phase                                                         
CLIMB             37730            37730             37730   37725   
CRUISE            45057            45057             45057   45057   
TAKEOFF           37890            37890             37890   37890   

              egt_probe_average  fuel_flw  corrected_fan_spd  core_spd  \
flight_phase                                                             
CLIMB                     37691     37684                  0     37696   
CRUISE                    45039     45039              42192     45057   
TAKEOFF                   37880     37880              37780     37890   

              zpn12p  vib_n1_#1_bearing  vib_n2_#1_bearing  \
flight_phase                                                 
CLIMB          37696              37539              37538   
CRUISE         45057              45008              45008   
TAKEOFF        37890              37832              37831   

              vib_n2_turbine_frame  eng_type  eng_number  
flight_phase                                              
CLIMB                        37539     37730       37730  
CRUISE                       45008     45057       45057  
TAKEOFF                      37832     37890       37890

### Initial Notes:  
1- We have 3 phases.  
2- From knowledge of the business domain, the critical phases(most stresses) that determine engine life are takeoff and climb.  
3- Also from bussiness domain, each takeoff is considered one running cycle that counts into the engine remaining useful life.  
4- Flight are ordered new to old (last flight of the engine is the first in top).  

In [16]:
# show the null percentage in each column
df.isnull().sum()/len(df)*100

eposition                0.000000
flight_phase             0.000000
flight_datetime          0.000000
eng_install_date         0.000000
cycles                   0.004143
egt_probe_average        0.055520
fuel_flw                 0.061321
corrected_fan_spd       33.730537
core_spd                 0.028174
zpn12p                   0.028174
vib_n1_#1_bearing        0.246940
vib_n2_#1_bearing        0.248597
vib_n2_turbine_frame     0.246940
eng_type                 0.000000
eng_number               0.000000
dtype: float64

In [17]:
# create two cloumns date and time from the flight_datetime, the current format is '25-12-2024  15:05:00 PM'
df['flight_datetime_c'] = pd.to_datetime(df['flight_datetime'], format='%d-%m-%y %H:%M', dayfirst=True)
df['date'] = df['flight_datetime_c'].dt.date
df['time'] = df['flight_datetime_c'].dt.time
df.head(30)


eposition flight_phase flight_datetime eng_install_date  cycles  \
0           1       CRUISE   29-11-23 8:56   24-08-23 23:00   334.0   
1           1       CRUISE   07-09-23 1:10   24-08-23 23:00    32.0   
2           1       CRUISE   19-01-24 3:09   24-08-23 23:00   532.0   
3           1        CLIMB  25-02-24 21:58   24-08-23 23:00   675.0   
4           1      TAKEOFF  10-07-24 16:28   24-08-23 23:00  1223.0   
5           1       CRUISE  23-02-25 23:45   24-08-23 23:00   126.0   
6           1       CRUISE   17-03-24 8:01   24-08-23 23:00   766.0   
7           1       CRUISE  08-12-24 12:53   24-08-23 23:00   118.0   
8           1      TAKEOFF   09-08-24 2:11   24-08-23 23:00    39.0   
9           1      TAKEOFF   30-06-24 9:25   24-08-23 23:00  1178.0   
10          1       CRUISE   03-10-23 3:45   24-08-23 23:00   126.0   
11          1      TAKEOFF   04-02-24 9:02   24-08-23 23:00   594.0   
12          1       CRUISE   19-12-24 8:32   24-08-23 23:00   154.0   
13          1        CLIMB  30-03-24 12:48   24-08-23 23:00   819.0   
14          1        CLIMB  05-06-24 14:13   24-08-23 23:00  1086.0   
15          1       CRUISE   01-12-24 2:30   24-08-23 23:00    96.0   
16          1       CRUISE   05-01-25 8:34   24-08-23 23:00   181.0   
17          1        CLIMB  30-11-24 13:51   24-08-23 23:00    94.0   
18          1       CRUISE   30-11-24 7:43   24-08-23 23:00    93.0   
19          1       CRUISE   12-01-24 5:08   24-08-23 23:00   504.0   
20          1       CRUISE  19-05-24 17:41   24-08-23 23:00  1021.0   
21          1      TAKEOFF  31-12-24 21:15   24-08-23 23:00   161.0   
22          1        CLIMB   25-02-24 5:48   24-08-23 23:00   672.0   
23          1      TAKEOFF  31-05-24 21:23   24-08-23 23:00  1062.0   
24          1        CLIMB   17-01-24 8:20   24-08-23 23:00   525.0   
25          1      TAKEOFF  30-04-24 18:55   24-08-23 23:00   944.0   
26          1       CRUISE  24-10-24 11:25   24-08-23 23:00   119.0   
27          1       CRUISE  09-07-24 17:00   24-08-23 23:00  1220.0   
28          1      TAKEOFF  01-02-25 14:10   24-08-23 23:00    55.0   
29          1        CLIMB   30-11-24 7:13   24-08-23 23:00    93.0   

    egt_probe_average  fuel_flw  corrected_fan_spd  core_spd  zpn12p  \
0            793.3125    2782.0            92.2129    103.12   86.67   
1            792.9250    2725.0            90.0819    103.99   86.02   
2            757.2250    2861.0            89.0560    101.87   83.67   
3            905.0375    5959.0                NaN    107.42   92.15   
4            912.3625    2911.0            85.3176    106.87   86.45   
5            806.4375    2650.0            92.0863    103.02   86.50   
6            715.6875    2504.0            88.0891     99.87   81.40   
7            786.9625    3005.0            92.2527    102.16   85.80   
8            983.2375    4066.0            86.9904    110.24   90.41   
9            938.3250    3206.0            85.0595    108.41   87.59   
10           801.7250    2964.0            89.9929    104.24   86.18   
11           883.4250    1917.0            84.8623    105.44   84.98   
12           792.7625    2851.0            91.5343    102.63   85.88   
13           880.2750    5963.0                NaN    106.23   91.36   
14           848.1250    5795.0                NaN    104.89   89.37   
15           790.1250    2910.0            89.2138    103.38   85.11   
16           737.3125    2801.0            89.1492    100.30   82.48   
17           880.3625    5631.0                NaN    105.81   90.54   
18           774.6000    3028.0            91.0355    102.02   84.99   
19           767.9000    2733.0            88.4549    102.49   83.74   
20           750.7875    2575.0            92.3709    100.53   84.52   
21           883.3375    7527.0            84.0543    105.25   84.36   
22           916.0000    6086.0                NaN    107.85   92.68   
23           894.4125    3734.0            80.9989    106.49   83.19   
24

In [18]:
#sort the data by engine number and flight datetime
df = df.sort_values(by=['eng_number', 'flight_datetime_c'])
df.head(30)

eposition flight_phase flight_datetime eng_install_date  cycles  \
105607          1      TAKEOFF   29-02-20 7:09   19-01-20 22:00    37.0   
102026          1        CLIMB   29-02-20 7:10   19-01-20 22:00    38.0   
100450          1       CRUISE   29-02-20 8:41   19-01-20 22:00    38.0   
104854          1       CRUISE   29-02-20 9:34   19-01-20 22:00    38.0   
104015          1      TAKEOFF  29-02-20 11:33   19-01-20 22:00    38.0   
108137          1        CLIMB  29-02-20 11:33   19-01-20 22:00    39.0   
104228          1       CRUISE  29-02-20 11:55   19-01-20 22:00    39.0   
107919          1       CRUISE  29-02-20 12:25   19-01-20 22:00    39.0   
99529           1        CLIMB  29-02-20 18:24   19-01-20 22:00    40.0   
104341          1      TAKEOFF  29-02-20 18:24   19-01-20 22:00    39.0   
107490          1       CRUISE  29-02-20 18:49   19-01-20 22:00    40.0   
99014           1      TAKEOFF  29-02-20 20:19   19-01-20 22:00    40.0   
103895          1        CLIMB  29-02-20 20:20   19-01-20 22:00    41.0   
98357           1       CRUISE  29-02-20 20:37   19-01-20 22:00    41.0   
102935          1        CLIMB   01-03-20 8:03   19-01-20 22:00    42.0   
103314          1      TAKEOFF   01-03-20 8:03   19-01-20 22:00    41.0   
96360           1       CRUISE   01-03-20 8:31   19-01-20 22:00    42.0   
98058           1        CLIMB  01-03-20 11:55   19-01-20 22:00    43.0   
106392          1      TAKEOFF  01-03-20 11:55   19-01-20 22:00    42.0   
104390          1       CRUISE  01-03-20 12:52   19-01-20 22:00    43.0   
99383           1        CLIMB  01-03-20 15:23   19-01-20 22:00    44.0   
101188          1      TAKEOFF  01-03-20 15:23   19-01-20 22:00    43.0   
101282          1       CRUISE  01-03-20 15:53   19-01-20 22:00    44.0   
97340           1      TAKEOFF  01-03-20 17:45   19-01-20 22:00    44.0   
105580          1        CLIMB  01-03-20 17:46   19-01-20 22:00    45.0   
95271           1       CRUISE  01-03-20 18:04   19-01-20 22:00    45.0   
95164           1      TAKEOFF  02-03-20 12:20   19-01-20 22:00    45.0   
98731           1        CLIMB  02-03-20 12:20   19-01-20 22:00    46.0   
100900          1       CRUISE  02-03-20 12:55   19-01-20 22:00    46.0   
103214          1      TAKEOFF  02-03-20 16:10   19-01-20 22:00    46.0   

        egt_probe_average  fuel_flw  corrected_fan_spd  core_spd  zpn12p  \
105607           810.5125    1520.0                NaN    103.19   79.21   
102026           851.5125    5373.0                NaN    105.20   89.15   
100450           754.8250    2431.0                NaN    102.94   83.01   
104854           577.1125    1749.0                NaN     94.61   53.87   
104015           876.5125    1476.0            82.9875    106.00   84.16   
108137           870.5875    5239.0                NaN    106.23   89.70   
104228           755.9250    2734.0            84.9793    103.48   82.65   
107919           744.6125    2687.0            84.5876    102.74   81.89   
99529            849.6125    5337.0                NaN    105.01   88.99   
104341           804.8500    1587.0            77.9703    102.90   78.20   
107490           722.0750    2894.0            81.8394     99.93   79.41   
99014            810.3625    2832.0            77.1413    103.33   77.98   
103895           866.2875    5439.0                NaN    105.87   89.71   
98357            722.2000    3023.0            81.1629    100.24   79.27   
102935           849.5625    5356.0                NaN    105.14   89.05   
103314           877.1875    1654.0            85.1409    105.55   85.20   
96360            767.7250    2414.0            91.2666    101.87   85.33   
98058            829.1875    5310.0                NaN    104.23   88.17   
106392           856.9375    1583.0            85.0521    104.80   84.43   
104390           734.0000    2182.0            89.8346    100.64   83.35   
99383            847.1625    5390.0                NaN    105.14   88.91   
101188    

In [19]:
# drop the flight_datetime column
df = df.drop('flight_datetime', axis=1)

In [20]:
# for each engine lets assign a flight cycle for each row, flight cycle start with take off and end with cruise
# flight cycle will increase by 1 for each take off from oldest to newest date for each engine
# first split daata by engine into a list of dataframes 
# then from oldest to newsest date for each engine assign a flight cycle for each row that increaes by 1 for each take off
# then combine the dataframes into one dataframe
df['flight_cycle'] = 0
df_list = [df[df['eng_number'] == i] for i in df['eng_number'].unique()]
for i in range(len(df_list)):
    df_list[i] = df_list[i].sort_values(by='flight_datetime_c')
    flight_cycle = 0
    for j in range(len(df_list[i])):
        if df_list[i].iloc[j]['flight_phase'] == 'TAKEOFF':
            flight_cycle += 1
        df_list[i].iloc[j, df_list[i].columns.get_loc('flight_cycle')] = flight_cycle

In [21]:
# combine the dataframes
df = pd.concat(df_list)

In [22]:
df.head(20)

eposition flight_phase eng_install_date  cycles  egt_probe_average  \
105607          1      TAKEOFF   19-01-20 22:00    37.0           810.5125   
102026          1        CLIMB   19-01-20 22:00    38.0           851.5125   
100450          1       CRUISE   19-01-20 22:00    38.0           754.8250   
104854          1       CRUISE   19-01-20 22:00    38.0           577.1125   
104015          1      TAKEOFF   19-01-20 22:00    38.0           876.5125   
108137          1        CLIMB   19-01-20 22:00    39.0           870.5875   
104228          1       CRUISE   19-01-20 22:00    39.0           755.9250   
107919          1       CRUISE   19-01-20 22:00    39.0           744.6125   
99529           1        CLIMB   19-01-20 22:00    40.0           849.6125   
104341          1      TAKEOFF   19-01-20 22:00    39.0           804.8500   
107490          1       CRUISE   19-01-20 22:00    40.0           722.0750   
99014           1      TAKEOFF   19-01-20 22:00    40.0           810.3625   
103895          1        CLIMB   19-01-20 22:00    41.0           866.2875   
98357           1       CRUISE   19-01-20 22:00    41.0           722.2000   
102935          1        CLIMB   19-01-20 22:00    42.0           849.5625   
103314          1      TAKEOFF   19-01-20 22:00    41.0           877.1875   
96360           1       CRUISE   19-01-20 22:00    42.0           767.7250   
98058           1        CLIMB   19-01-20 22:00    43.0           829.1875   
106392          1      TAKEOFF   19-01-20 22:00    42.0           856.9375   
104390          1       CRUISE   19-01-20 22:00    43.0           734.0000   

        fuel_flw  corrected_fan_spd  core_spd  zpn12p  vib_n1_#1_bearing  \
105607    1520.0                NaN    103.19   79.21             0.1000   
102026    5373.0                NaN    105.20   89.15             0.1267   
100450    2431.0                NaN    102.94   83.01             0.1133   
104854    1749.0                NaN     94.61   53.87             0.0733   
104015    1476.0            82.9875    106.00   84.16             0.0600   
108137    5239.0                NaN    106.23   89.70             0.1267   
104228    2734.0            84.9793    103.48   82.65             0.1133   
107919    2687.0            84.5876    102.74   81.89             0.1333   
99529     5337.0                NaN    105.01   88.99             0.1267   
104341    1587.0            77.9703    102.90   78.20             0.1133   
107490    2894.0            81.8394     99.93   79.41             0.0600   
99014     2832.0            77.1413    103.33   77.98             0.0333   
103895    5439.0                NaN    105.87   89.71             0.0867   
98357     3023.0            81.1629    100.24   79.27             0.1467   
102935    5356.0                NaN    105.14   89.05             0.1733   
103314    1654.0            85.1409    105.55   85.20             0.0867   
96360     2414.0            91.2666    101.87   85.33             0.1667   
98058     5310.0                NaN    104.23   88.17             0.1867   
106392    1583.0            85.0521    104.80   84.43             0.0867   
104390    2182.0            89.8346    100.64   83.35             0.1733   

        vib_n2_#1_bearing  vib_n2_turbine_frame eng_type eng_number  \
105607              0.275                 0.700        2          1   
102026              0.375                 0.500        2          1   
100450              0.425                 0.550        2          1   
104854              0.050                 0.850        2          1   
104015              0.225                 0.600        2          1   
108137              0.150                 0.450        2          1   
104228              0.225                 0.475        2          1   
107919              0.175                 0.450        2          1   
99529               0.400                 0.375        2          1   
104341              0.300                 0.750        2      

In [23]:
# add a column for remaining useful life (RUL) for each engine
# RUL is the number of flight cycles remaining for the engine before it fails
# RUL = max(flight_cycle for this engine) - flight_cycle
# store the max flight cycle for each engine in a dictionary
max_flight_cycle = {}
for i in df['eng_number'].unique():
    max_flight_cycle[i] = df[df['eng_number'] == i]['flight_cycle'].max()
df['RUL'] = 0
for i in range(len(df)):
    df.iloc[i, df.columns.get_loc('RUL')] = max_flight_cycle[df.iloc[i]['eng_number']] - df.iloc[i]['flight_cycle']


In [24]:
# lets check the data
df.head(20)

eposition flight_phase eng_install_date  cycles  egt_probe_average  \
105607          1      TAKEOFF   19-01-20 22:00    37.0           810.5125   
102026          1        CLIMB   19-01-20 22:00    38.0           851.5125   
100450          1       CRUISE   19-01-20 22:00    38.0           754.8250   
104854          1       CRUISE   19-01-20 22:00    38.0           577.1125   
104015          1      TAKEOFF   19-01-20 22:00    38.0           876.5125   
108137          1        CLIMB   19-01-20 22:00    39.0           870.5875   
104228          1       CRUISE   19-01-20 22:00    39.0           755.9250   
107919          1       CRUISE   19-01-20 22:00    39.0           744.6125   
99529           1        CLIMB   19-01-20 22:00    40.0           849.6125   
104341          1      TAKEOFF   19-01-20 22:00    39.0           804.8500   
107490          1       CRUISE   19-01-20 22:00    40.0           722.0750   
99014           1      TAKEOFF   19-01-20 22:00    40.0           810.3625   
103895          1        CLIMB   19-01-20 22:00    41.0           866.2875   
98357           1       CRUISE   19-01-20 22:00    41.0           722.2000   
102935          1        CLIMB   19-01-20 22:00    42.0           849.5625   
103314          1      TAKEOFF   19-01-20 22:00    41.0           877.1875   
96360           1       CRUISE   19-01-20 22:00    42.0           767.7250   
98058           1        CLIMB   19-01-20 22:00    43.0           829.1875   
106392          1      TAKEOFF   19-01-20 22:00    42.0           856.9375   
104390          1       CRUISE   19-01-20 22:00    43.0           734.0000   

        fuel_flw  corrected_fan_spd  core_spd  zpn12p  vib_n1_#1_bearing  \
105607    1520.0                NaN    103.19   79.21             0.1000   
102026    5373.0                NaN    105.20   89.15             0.1267   
100450    2431.0                NaN    102.94   83.01             0.1133   
104854    1749.0                NaN     94.61   53.87             0.0733   
104015    1476.0            82.9875    106.00   84.16             0.0600   
108137    5239.0                NaN    106.23   89.70             0.1267   
104228    2734.0            84.9793    103.48   82.65             0.1133   
107919    2687.0            84.5876    102.74   81.89             0.1333   
99529     5337.0                NaN    105.01   88.99             0.1267   
104341    1587.0            77.9703    102.90   78.20             0.1133   
107490    2894.0            81.8394     99.93   79.41             0.0600   
99014     2832.0            77.1413    103.33   77.98             0.0333   
103895    5439.0                NaN    105.87   89.71             0.0867   
98357     3023.0            81.1629    100.24   79.27             0.1467   
102935    5356.0                NaN    105.14   89.05             0.1733   
103314    1654.0            85.1409    105.55   85.20             0.0867   
96360     2414.0            91.2666    101.87   85.33             0.1667   
98058     5310.0                NaN    104.23   88.17             0.1867   
106392    1583.0            85.0521    104.80   84.43             0.0867   
104390    2182.0            89.8346    100.64   83.35             0.1733   

        vib_n2_#1_bearing  vib_n2_turbine_frame eng_type eng_number  \
105607              0.275                 0.700        2          1   
102026              0.375                 0.500        2          1   
100450              0.425                 0.550        2          1   
104854              0.050                 0.850        2          1   
104015              0.225                 0.600        2          1   
108137              0.150                 0.450        2          1   
104228              0.225                 0.475        2          1   
107919              0.175                 0.450        2          1   
99529               0.400                 0.375        2          1   
104341              0.300                 0.750        2      

In [25]:
max_flight_cycle

{1: 4429,
 2: 4690,
 3: 4510,
 4: 4035,
 5: 4427,
 6: 4089,
 7: 3975,
 8: 2018,
 9: 1888,
 10: 1967,
 11: 1862}

In [26]:
# lets add  max_flight_cycle dict to the engine life dict in days
engine_life = {}
for i in df['eng_number'].unique():
    engine_life[i] = (df[df['eng_number'] == i]['date'].max() - df[df['eng_number'] == i]['date'].min()).days

# Add max_flight_cycle to engine_life
for i in engine_life.keys():
    engine_life[i] = {'Life in Days': engine_life[i], 'Life in FC': max_flight_cycle[i]}
engine_life

{1: {'Life in Days': 1736, 'Life in FC': 4429},
 2: {'Life in Days': 1648, 'Life in FC': 4690},
 3: {'Life in Days': 1830, 'Life in FC': 4510},
 4: {'Life in Days': 1797, 'Life in FC': 4035},
 5: {'Life in Days': 1715, 'Life in FC': 4427},
 6: {'Life in Days': 1536, 'Life in FC': 4089},
 7: {'Life in Days': 1468, 'Life in FC': 3975},
 8: {'Life in Days': 646, 'Life in FC': 2018},
 9: {'Life in Days': 533, 'Life in FC': 1888},
 10: {'Life in Days': 583, 'Life in FC': 1967},
 11: {'Life in Days': 564, 'Life in FC': 1862}}

In [30]:
# save the cleaned data to a csv file
df.to_csv('engines_data_cleaned.csv', index=False)


### Notes:  
1- Our 11 engines life time ranges from 1862 FC to 4690 FC, from 533 days to 1830 days. 

2- Significant difference in life between type 1 Engines (shorter life) and type 2 engines. 

3- We have the following features to analyze: flight_cycle, egt_probe_average, fuel_flw, corrected_fan_spd, core_spd, zpn12p, vib_n1_#1_bearing, vib_n2_#1_bearing, vib_n2_turbine_frame, RUL   

4- The features are for 3 phases on 11 engines.   